src/ivalue_manager.rs

```diff
@@ -63,11 +68,81 @@ fn update<F: FnMut(IValue) -> Result<Option<IValue>, Error>>(
                 if let Ok(x) = token.parse::<usize>() {
                     if is_last {
                         if x < arr.len() {
-                            let v = std::mem::replace(&mut arr[x], IValue::NULL);
-                            if let Some(res) = (func)(v)? {
-                                arr[x] = res;
-                            } else {
-                                arr.remove(x);
+                            let mut v = &mut arr.as_mut_slice()[x];
+                            match (func)(&mut v) {
+                                Ok(res) => {
+                                    if let Some(res) = res {
+                                        *v = res;
+                                    } else {
+                                        arr.remove(x);
+                                    }
+                                }
+                                Err(err) => return Err(err),
+                            }
+                        }
+                        return Ok(());
+                    }
+                    arr.get_mut(x)
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        };
+
+        if let Some(t) = target_opt {
+            target = t;
+        } else {
+            break;
+        }
+    }
+
+    Ok(())
+}
+
+fn update_in_closure<F: FnMut(&mut IValue) -> Result<Option<()>, Error>>(
```

Comment:
Can you add a documentation to this function that explain where is it used and how is it different from the one above?

Reply:
Added explanation comment in `fn update`

Reply:
Don't see it

Reply:
Will add in code (not only in GitHub comment)